[3 - 4단계 방탈출 예약 대기] 밍트(김명지) 미션 제출합니다.

Starlight258 opened the pull request:
파랑 안녕하세요~ 밍트입니다! 😁

이번 미션 잘 부탁드립니다~
감사합니다 🙇‍♀️
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  [작성한 프롤로그 링크](https://prolog.techcourse.co.kr/studylogs/4350)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

- [x] 규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.
- [x] 규칙 2: else 예약어를 쓰지 않는다.
- [-] 규칙 3: 모든 원시값과 문자열을 포장한다.
- [x] 규칙 4: 한 줄에 점을 하나만 찍는다.
- [x] 규칙 5: 줄여쓰지 않는다(축약 금지).
- [x] 규칙 6: 모든 엔티티를 작게 유지한다.
- [-] 규칙 7: 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
- [-] 규칙 8: 일급 콜렉션을 쓴다.
- [x] 규칙 9: 게터/세터/프로퍼티를 쓰지 않는다.
  - 응답(json, view)에서만 사용 

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
3개 이상의 인스턴스 변수로 나눌만큼 객체의 크기가 크지 않았고, 일급 컬렉션을 쓸만한 객체가 보이지 않았습니다.
요청 파라미터, 상수 값을 제외하고 모든 원시값과 문자열을 포장하였습니다.

## 구현 방식
### 예약 대기 방식
- 예약 대기시 `자동 승인`되도록 구현하였습니다.
  - 예약 취소시 하나하나 승인 버튼을 누르는 것이 예약이 많아짐에 따라 불편함을 초래할 수 있기 때문입니다. 

### 예약 확정 인원 판단 방식
- 별도의 확정 회원 컬럼을 두지 않고 `예약 생성 시간`을 기준으로 확정 회원을 판단하였습니다.
- 확정된 회원이 취소할 경우, 가장 먼저 대기한 회원이 자동으로 확정 상태로 승격됩니다.

### 테이블 설계 (다대다 + 중간 엔티티)

![image](https://github.com/user-attachments/assets/91fc317e-00c3-4b8f-af20-74f9a868b7b8)

- 예약 슬롯(`ReservationSlot`): 예약된 슬롯 (날짜 + 시간 + 테마)의 단위입니다.
- 예약(`Reservation`) :  회원이 특정 예약 슬롯에 예약을 신청(확정 or 대기)한 기록을 나타내는 중간 엔티티입니다.

- 예약 슬롯과 회원은 `다대다` 관계입니다.
  - 하나의 예약 슬롯은 여러 회원이 예약할 수 있습니다.
  - 한 회원은 다른 여러 예약 슬롯을 예약할 수 있습니다.

- `다대다` 관계이므로, 중간 엔티티인 `예약` 테이블을 두어 확장성을 고려하였습니다.
  -  중간 엔티티인 예약에서는 예약 상태를 관리하고, 순서를 관리할 수 있습니다.
  - 중간 테이블을 두어 추가 정보에 대한 컬럼을 추가할 수 있는 장점이 있습니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

객체지향적으로 잘 설계되었는지, 테이블이 잘 설계되었는지 피드백 주시면 감사하겠습니다.
추가로 jpa를 잘 활용하여 미션을 잘 구현하였는지 봐주시면 감사하겠습니다!

Starlight258 committed:
feat: 예약 대기 프론트엔드 코드 추가

Starlight258 committed:
feat: data.sql에 추가된 필드 반영

Starlight258 committed:
feat: 예약 대기 요청 기능 구현

Starlight258 committed:
feat: 멤버 예약 목록 조회시 대기 목록 조회 기능 구현

Starlight258 committed:
chore: 데이터베이스 이름 수정

Starlight258 committed:
docs: 기능 목록 수정

Starlight258 committed:
feat: 예약 대기 취소 기능 구현

Starlight258 committed:
docs: 주석 제거

Starlight258 committed:
feat: 관리자용 예약 대기 관리 페이지 추가

Starlight258 committed:
feat: 관리자용 예약 대기 목록 조회 기능 구현

Starlight258 committed:
feat: 예약 대기 취소 기능 구현

Starlight258 committed:
feat: db 컬럼명 수정

- 테이블명 복수형으로 수정
- id는 테이블마다 명시적으로 테이블명 포함하여 지정

Starlight258 committed:
feat: reservation -> bookingSlot으로 엔티티명 변경

- 특정 날짜, 시간, 테마의 예약 슬롯이므로 bookingSlot으로 엔티티명 변경

Starlight258 committed:
refactor: Waiting -> Reservation으로 엔티티명 변경

- Waiting은 예약한 회원도 포함되므로 직관적이지 않다.

Starlight258 committed:
feat: BookingSlot -> ReservationSlot으로 엔티티명 변경

- bookingSlot은 예약 가능한 slot을 암시한다.
- 예약된 slot만 표시되므로 명확하게 엔티티명을 변경한다.

Starlight258 commented on src/main/java/roomescape/reservation/presentation/ReservationController.java:
```diff
@@ -30,53 +26,41 @@ public ReservationController(final ReservationApplicationService reservationAppl
-            @PathVariable("id") Long id
+    @RequireRole(MemberRole.ADMIN)
+    @DeleteMapping("/admin/waiting-reservations/{waitingId}")
+    public ResponseEntity<Void> removeWaitingReservation(
```
## 예약 취소 상태 관리
- 현재 예약 대기 중인 회원이 예약 대기를 취소할 때, 삭제하는 방식으로 구현하였습니다.
- 취소 상태를 두어서 관리하는 것이 나을지, 현재 방식을 유지(삭제)하는 것이 좋을지 궁금합니다.

summerlunaa replied:
> 취소 상태를 두어서 관리하는 것이 나을지, 현재 방식을 유지(삭제)하는 것이 좋을지 궁금합니다.

각각의 방식에는 어떤 장단점이 있을까요? 현재는 어떤 방식이 더 어울릴까요? 미래에 어떤 요구사항이 추가될 수 있을까요? 미래의 요구사항을 고려했을 때는 또 어떤 방식이 적절할까요? 혹은 취소 상태를 두어 관리하다 삭제하는 방식으로 넘어가는 비용과 그 반대로 변경하는 비용은 어떻게 될까요? 이런 질문들을 스스로 던져보고 정리해보세요 :)

Starlight258 replied:
## 취소 상태 방식 
### 장/단점
- 취소 상태를 둘 경우에 취소한 예약을 관리할 수 있습니다. 
- 단점으로는 취소된 예약이 db에 저장되어 저장할 데이터가 많아지는 문제입니다.

### 변경 비용
- 변경 비용은 적습니다. `soft delete`를 하고, 단순히 `삭제 필드`와 `예약 상태 필드`(확정, 대기, 취소)만 추가하면 되기 때문입니다. 

### 결론
현재는 삭제 방식을 유지하는 것이 좋을 것 같습니다. 후에 요구사항이 추가된다면 충분히 쉽고 빠르게 변경할 수 있고, 취소 여부를 서버 로그를 통해 확인할 수 있기 때문입니다.

summerlunaa replied:
넵넵 좋습니다! 하지만 시간이 많이 흘러 취소된 데이터가 많아지면 로그를 통해 하나하나 취소된 예약 정보를 확인인해서 다시 데이터를 넣어주기는 굉자히 어려울 것 같긴 합니다 :) 추후 예약 취소 내역 조회 기능이 추가될 가능성이 높다면 우선 취소 상태를 사용하는 방향으로 하되, 기능이 불필요해지면 데이터를 전부 제거하는 방식도 괜찮겠네요. 다만 현재 상황에서는 그럴 가능성이 없으니 데이터를 삭제하는 편이 더 알맞아 보입니다.

Starlight258 commented on src/main/java/roomescape/reservationslot/application/ReservationSlotApplicationService.java:
```diff
@@ -0,0 +1,75 @@
+import roomescape.theme.domain.service.ThemeDomainService;
+import roomescape.reservation.domain.service.ReservationDomainService;
+
+@Service
+public class ReservationSlotApplicationService {
```
## `@Transactional`
Spring Data JPA가 제공하는 메서드는 기본적으로 `@Transactional`을 가지고 있기 때문에 서비스에 `@Transactional`을 붙이지 않았습니다.
그리고 데이터를 업데이트하는 jpql를 사용할 경우에만(`@Modifying`) `@Transactional`을 붙여주었습니다.

지금까지는 이 방식으로 잘 동작했지만, 서비스 메서드 단위로 트랜잭션을 묶어주면 데이터 일관성 측면에서 더 좋을지 고민중입니다.

질문 
- `@Transactional`을 붙이고, 붙이지 않는 기준이 있나요?
- 현재 `ApplicationService` -> `DomainService`로 서비스 계층을 나누었는데, `@Transactional`을 붙인다면 어느쪽이 좋을까요?
-  조회만 하는 메서드에도 `@Transactional(readOnly=true)`를 붙이는 것이 좋을까요?

summerlunaa replied:
좋은 고민입니다! 제가 답을 드리기보다는 답을 찾기 위한 질문들을 드리는 게 더 도움이 될 것 같아요

- `@Transactional`을 붙이고, 붙이지 않는 기준이 있나요?
   - `@Transactional`은 어떤 역할을 하나요?
   - `@Transactional`은 언제 필요한가요?
- 현재 `ApplicationService` -> `DomainService`로 서비스 계층을 나누었는데, `@Transactional`을 붙인다면 어느쪽이 좋을까요?
   - 지금은 어디에 붙여도 상관이 없어 보이겠지만 다음 미션인 `결제` 기능을 도입한다고 가정해보세요. 결제 기능을 도입하면 외부 API를 호출해야 할텐데 외부 API 호출은 어디에서 이루어지는 게 좋을까요? 외부 API 호출과 트랜잭션은 어떤 연관성이 있을까요?
- 조회만 하는 메서드에도 `@Transactional(readOnly=true)`를 붙이는 것이 좋을까요?
   - 밍트는 어떻게 생각하나요? 의견을 먼저 얘기해주시면 더 좋을 것 같아용!!!

Starlight258 replied:
공부할 수 있는 질문 주셔서 감사합니다 🙇‍♀️

## `@Transactional`
### 역할
- `@Transactional`은 `ACID`를 보장한다.
- 또한 트랜잭션의 경계를 명시적으로 설정하여 어디서 시작하고 끝날지 결정한다.

### 필요성 
- 데이터 변경 작업(`INSERT`, `UPDATE`, `DELETE`)을 하는 경우
  - 더티 체킹이 트랜잭션 내에서만 동작하므로 데이터 변경 사항을 db에 반영하려면 트랜잭션이 필요하다.
- 여러 테이블간 데이터 일관성이 보장되어야하는 경우
- 비즈니스 로직 실행 중 예외 발생시 실행 로직 전부가 롤백되어야할 경우 (데이터 무결성이 보장되어야하는 경우)

### 불필요한 경우
- 단순 조회만 수행하는 경우
  - 데이터 변경이 없으므로 X
- 계산이나 변환 로직만 수행하거나, 캐시나 정적 데이터만 접근하는 경우
  - db 접근이 없으므로  X

### 결제 기능: 서비스 계층에서 `@Transactional`을 붙이는 곳
>  `ApplicationService`<-`DataService`

- 결제 기능을 도입한다면, `Client` 객체가 외부 API 호출을 수행하도록 구현할 것입니다. 즉, `ApplicationReservationService`는 `DataService`와 `Client`를 의존합니다.
- 결제 api는 `외부 api`이므로 db와 관련되어 있지 않은 작업입니다. (예약과 함께 트랜잭션으로 묶는다고 요청 보내는 것이 `rollback`되지 않습니다). 또한 트랜잭션으로 묶은 경우 `외부 api`가 지연되면 `db 커넥션`을 오랫동안 점유하는 문제가 있습니다. 따라서 `ApplicationService` 클래스에 `@Transactional`을 붙이지 않고, 예약과 결제를 수행하는 메서드를 분리하여 예약하는 부분만 `@Transactional`을 붙일 것 같습니다.

```java
class ApplicationReservationService {

  private final ReervationDataService reservationDataService;
  private final Client client;

  public ReservationResponse createReservationWithPayment(){ // 예약 + 결제
       createReservation();
       payReservation();
  }
 
  @Transactional
  public ReservationResponse createReservation() { // 예약
  }

  public ReservationResponse payReservation() { // 결제  
}
```

## `@Transactional(readOnly=true)`
- 데이터 변경 작업을 하지 않고, 읽기 작업만 수행함을 명시적으로 선언합니다.
- 더티 체킹이 비활성화되어 성능상 장점이 있습니다.
- 데이터 변경 코드를 추가하는 실수를 방지할 수 있습니다.

### vs `@Transctional `안붙이는 경우
- 해당 어노테이션을 안붙일 경우 기본으로 `autoCommit` 설정이 되어 각 쿼리마다 별도 커넥션과 트랜잭션이 수행됩니다.
- `복잡한 쿼리 조회`에서는 쿼리 간 일관성을 보장할 수 없습니다.
- `대량 데이터일` 경우 더티 체킹 비활성화 효과를 위해 붙이기도 합니다.

### vs `@Transctional(readOnly = false)` 붙이는 경우
- 조회만 하는데 스냅샷을 저장하고 더티 체킹을 수행하는 등의 오버헤드가 있습니다.

### 결론
- 단순 조회만 할 경우 어노테이션을 붙이거나 `readOnly=true`로 붙이거나 큰 차이가 없는 것 같습니다.
- 복잡한 쿼리 조회나 대량 데이터는 `@Transactional(readOnly=true)`를 붙입니다.

[feat: `@transactional` 어노테이션 서비스에 추가아혀 일관성을 높임](https://github.com/Starlight258/spring-roomescape-waiting/commit/4f719a20cb4430f17770b40aa9bb4b822410e074)

summerlunaa replied:
너무 잘 정리해주셨네요 👍 데이터 변경을 막기 위해 단순 조회에서도 `readOnly=true`를 붙이기도 하고, 추후 쓰기 전용 db와 읽기 전용 db가 분리되면 조회의 경우 읽기 전용 db를 사용하도록 하기 위해 `readOnly=true`를 명시해주기도 합니다. 참고차 말씀드려요 :)

Starlight258 committed:
refactor: 사용하지 않은 코드 제거

Starlight258 committed:
feat: 동적 필터링 jpql로 구현

Starlight258 committed:
fix: 연관관계 매핑시 @JoinColumn 사용하도록 수정

Starlight258 committed:
fix: 외래키 주인이 아닐 경우 @JoinColumn 사용하지 않도록 수정

Starlight258 committed:
refactor: 적절히 클래스명 변경

summerlunaa commented on src/main/java/roomescape/reservation/domain/ReservationStatus.java:
```diff
@@ -2,15 +2,15 @@
 public enum ReservationStatus {
 
-    RESERVED("예약"), WAITING("예약대기");
+    CURRENT("예약"), WAITING("%d번째 예약대기");
```
참고) 가능하면 서버에서는 대기 순서 숫자와 예약 상태만 내려주고 사용자에게 보여지는 문자열은 view(프론트)에서 만들어주는 게 좋긴 합니다! 지금은 주어진 js를 사용해야하니 어쩔 수 없지만요 ㅎㅎ

Starlight258 replied:
답변 감사합니다! 프론트에서 수행해야하는 코드였군요!

summerlunaa commented on src/main/java/roomescape/reservation/exception/ReservationOwnerException.java:
```diff
@@ -0,0 +1,8 @@
+package roomescape.reservation.exception;
+
+public class ReservationOwnerException extends RuntimeException {
```
예외가 추가되었네욥! GlobalExceptionHandler에도 추가해주어야겠죠? 전체적으로 빠진 예외들 점검해보세요~

Starlight258 replied:
해당 예외는 더이상 사용되지 않아 제거하였습니다! 그리고 전체적으로 빠진 예외를 확인하여 추가하였습니다!

[refactor: 사용하지 않은 클래스 삭제](https://github.com/Starlight258/spring-roomescape-waiting/commit/da0870a3bfcd1bf817b4d146fc3de5a6811ca7b9)

summerlunaa commented on src/main/java/roomescape/reservation/application/ReservationApplicationService.java:
```diff
@@ -1,69 +1,53 @@
-    }
+    public ReservationResponse addWaiting(final LocalDate date, final Long timeId, final Long themeId,
+                                          final Long memberId) {
+        ReservationSlot reservationSlot = reservationSlotDomainService.getReservationByDateAndTimeAndTheme(date, timeId, themeId);
```
기존의 reservation 도메인을 reservation slot과, reservation으로 나누셨네요! 두 개념이 되게 비슷해서 헷갈릴 수 있는데 잘 구분해서 분리하신 것 같아요! 좋습니다 👍 
여기서는 reservation slot을 조회하고 있는 걸로 보이는데요, 메서드명은 `getReservationXXX..`이네요! 두 개념이 헷갈리지 않도록 메서드명을 명확하게 변경하는 게 좋아보여요. 전체적으로 점검해보세요!

Starlight258 replied:
먼저 코드를 보셨을 때 네이밍으로 인해 혼란을 드린점 죄송합니다 🙇‍♀️
설계를 변경하면서 클래스명과 메서드명이 미처 다 바뀌지 않고 pr 제출됨을 확인하였습니다 😭
미처 바뀌지 않은 메서드명과 변수명 변경하였습니다! 
앞으로는 pr 제출 전에 한번 더 코드를 확인하겠습니다.

summerlunaa commented on src/main/java/roomescape/reservation/application/ReservationApplicationService.java:
```diff
@@ -1,69 +1,53 @@
-        Member member = memberDomainService.getMember(memberId);
+    public void removeWaiting(final Long reservationId, final Long memberId) {
+        reservationDomainService.deleteByReservationSlotIdAndMemberId(reservationId, memberId);
+    }
```
여기도 reservationId가 아니라 reservationSlotId가 맞는거죠?

Starlight258 replied:
명확하게 수정하였습니다! 대규모로 리팩토링하다보니 미처 바뀌지 않은 이름들이 존재하는 것 같습니다..

summerlunaa commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -10,103 +10,79 @@
-
     @Enumerated(EnumType.STRING)
-    @Column(name = "reservation_status", nullable = false)
+    @Column(name = "waiting_status", nullable = false)
```
왜 컬럼명이 waiting status로 변경되었나요~? 아래도 reservation slot의 컬럼명은 booking slot id네요 이름을 다르게 설정한 이유가 궁금합니다.

Starlight258 replied:
`ReservationSlot`의 이전 이름이 `BookingSlot`입니다. 이름이 변경되어야하는 것이 맞습니다 🤧 수정했습니다 !

이름을 변경한 이유는 확정된 예약이 존재하는 경우에만 `Slot`이 존재하므로 `BookingSlot`(예약 가능한 슬롯)보다는 `ReservationSlot`(예약된 슬롯)이 더 적합하기 때문입니다.

현재 `status` 필드는 제거된 상태입니다! `ReservationSlot`의 id를 참조하는 예약들 중, 가장 첫번째 예약을 확정 예약으로 판단하여 계산합니다. 이를 통해 `status`와 실제 예약 데이터의 불일치 가능성을 제거했습니다.

summerlunaa commented on src/main/java/roomescape/reservation/infrastructure/JpaReservationRepository.java:
```diff
@@ -1,49 +1,41 @@
+    @Query("""
+            DELETE FROM Reservation w 
+            WHERE w.reservationSlot.id = :reservationId
+            AND w.member.id = :memberId
```
DELETE 쿼리도 직접 작성해주셨네요~? 불필요한 select 쿼리를 피하기 위함일까요?

summerlunaa replied:
이전 PR의 쿼리 관련 코멘트에도 답변 드렸어요! https://github.com/woowacourse/spring-roomescape-waiting/pull/249#discussion_r2107451049

Starlight258 replied:
해당 메서드는 굳이 쿼리를 작성하기보다, `spring data jpa`의 쿼리 메서드를 사용하는 것이 좋을 것 같습니다.
`jpql`로 작성할 만큼 복잡한 쿼리도 아니고, `fetch join`할 경우도 아니기 때문입니다.
따라서 가독성 측면에서 `쿼리 메서드`를 작성하도록 수정하겠습니다!

summerlunaa commented on src/main/java/roomescape/reservation/presentation/ReservationController.java:
```diff
@@ -30,53 +25,41 @@ public ReservationController(final ReservationApplicationService reservationAppl
-    @PostMapping("/admin/reservations")
-    public ResponseEntity<ReservationResponse> createReservation(
-            @RequestBody AdminReservationCreateRequest request
+    @GetMapping("/admin/waiting-reservations")
```
어드민은 컨트롤러를 분리하는 게 path나 RequireRole 관리에 있어서 더 편리할 수도 있겠네요.

Starlight258 replied:
맞습니다! `controller`에 권한별로 분리해서 관리하면 훨씬 파악하기 좋을 것 같습니다.
[refactor: 권한별 controller 분리](https://github.com/Starlight258/spring-roomescape-waiting/commit/a6fc40337807a4fcf9f4743df967cb7cccb37126)

summerlunaa commented on src/main/java/roomescape/reservationslot/application/ReservationSlotApplicationService.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public ReservationSlotResponse create(final LocalDate date, final Long timeId, final Long themeId, final Long memberId,
+                                          final LocalDateTime now) {
```
```suggestion
                                          final LocalDateTime reservationDateTime) {
```
now 보다는 예약 일시라는 의미의 필드명이 더 적절해보여요.

Starlight258 replied:
더 적합한 필드명이네요! 수정하였습니다.
[refactor: now -> reservationDateTime으로 변경](https://github.com/Starlight258/spring-roomescape-waiting/commit/0b9c8d222034c1b16edf2c26d1221fbf847b349e)

summerlunaa commented on src/main/java/roomescape/reservationslot/application/ReservationSlotApplicationService.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public List<MyReservationSlotResponse> findMyReservations(final MemberInfo memberInfo) {
+        return reservationDomainService.findMyReservations(memberInfo);
```
여기도 서비스에서는 `My`라는 상대적인 단어보다는 회원 ID로 예약을 찾는다는 의미에서 `findReservationsByMemberId`가 더 적절하겠네요.

Starlight258 replied:
맞습니다! `My`라는 상대적인 단어보다는 구체적이고 직관적으로 이해할 수 있는 네이밍을 짓도록 노력하겠습니다! 🫡

[feat: api endpoint 수정 및 메서드명 변경](https://github.com/Starlight258/spring-roomescape-waiting/commit/cc3b385e9ea8dfd3c0d63c658ad508c9c8d92962)

summerlunaa commented on src/main/java/roomescape/reservationslot/domain/ReservationSlot.java:
```diff
@@ -0,0 +1,140 @@
+        }
+        return Objects.equals(getId(), that.getId()) && Objects.equals(getDate(), that.getDate())
+                && Objects.equals(getTime(), that.getTime()) && Objects.equals(getTheme(),
+                that.getTheme()) && Objects.equals(reservations, that.reservations);
```
reservation에서는 id로만 동등성을 비교하는데 slot에서는 또 다르네요. 이유가 있을까요?

Starlight258 replied:
기본으로 생성된 `equals` & `hashCode`를 사용하였는데, 엔티티는 `id`를 기반으로 동등성을 판단하므로 수정하는게 좋을 것 같습니다! 반영했습니다!

Starlight258 committed:
chore: 데이터 정합성 맞도록 수정

summerlunaa commented on src/main/java/roomescape/reservationslot/domain/ReservationSlot.java:
```diff
@@ -0,0 +1,140 @@
+                .stream()
+                .sorted(Comparator.comparing(Reservation::getCreatedAt))
+                .toList();
+        return sortedReservations.indexOf(givenReservation);
```
`sorted`를 사용해도 원본 데이터는 변경되지 않아서 굳이 arrayList를 새로 생성할 필요는 없겠네요. 그리고 waiting을 추가할 때 매번 순차적으로 데이터를 넣는다고 가정하면 조회할 때마다 정렬을 할 필요는 없어 보이는데 어떻게 생각하시나요~?
추가로 rank는 0번부터 시작되나요?

Starlight258 replied:
스트림은 원본 컬렉션이 변경되지 않는군요! 😁

순차적으로 데이터를 넣긴 하지만, 만약 `데이터 저장 시점`과 `예약 생성 시점`이 다를 경우 문제가 생길 것 같아요.
예를 들어 서버상의 문제로 일부 서버의 예약이 불가능해서 이후에 보완 조치로 예약을 생성하는 경우가 있다고 가정해보겠습니다.
`id`가 더 커도 `예약 생성 일자`는 다른 예약보다 더 빠른 예약이 존재할 것 같습니다. 따라서 위험하기 때문에 `createdAt`을 두었습니다.

`rank`는 0부터 시작합니다. `rank`는 예약 대기 번째로, 첫번째 예약은 `예약 확정`이므로 기다리는 대기가 0입니다.
두번째 예약은 `1`명만 대기하면 되므로 1이 반환됩니다.

[feat: @orderby를 정의하여 컬렉션 로드시 자동으로 정렬되는 기능 추가](https://github.com/Starlight258/spring-roomescape-waiting/commit/74a226453615a7c40eeba1d04e8b3028df8703d2)

summerlunaa replied:
아하 설명 감사합니다 ㅎㅎ 보완 조치로 예약을 생성할 때 createdAt도 더 앞선 시간으로 설정해서 넣어줄 수도 있다는 거죠? 좋습니다!

> rank는 0부터 시작합니다. rank는 예약 대기 번째로, 첫번째 예약은 예약 확정이므로 기다리는 대기가 0입니다.
두번째 예약은 1명만 대기하면 되므로 1이 반환됩니다.

이 부분도 설명 들으니 이해가 되었습니다 👍

Starlight258 committed:
refactor: 컬럼 및 메서드명 수정

summerlunaa commented on src/test/java/roomescape/reservationslot/infrastructure/JpaReservationSlotRepositoryTest.java:
전체적으로 테스트가 부족해보이고 패키지도 업데이트가 안 된 것 같네요 😢 우선 화요일부터 다음 미션 시작이니 당장 무리해서 테스트를 채울 필요는 없지만 다음부터는 가능하면 TDD로 챙겨 보세요~

Starlight258 replied:
전반적으로 부족한 테스트를 보완하고 패키지 구조를 수정하였습니다!
항상 기본기에 충실한 개발자가 되기 위해 노력하겠습니다! 감사합니다.

summerlunaa reviewed (changes requested):
밍트 안녕하세요!
3-4단계도 잘 구현해주셨네요! reservation과 slot을 잘 분리해서 설계하신 것 같아요. 리뷰 확인 부탁드립니다~ 다음 미션 진행으로 인해 반영이 어려운 부분이 있으면 말씀해주세요.
이번에도 궁금한 점 있으시면 편하게 질문하시구요 :)

Starlight258 committed:
refactor: 불필요한 adapter 패턴 제거

Starlight258 committed:
feat: 불필요한 exist 쿼리 제거

Starlight258 committed:
refactor: DomainService -> DataService로 변경

- repository 위의 추상화 계층을 제공하는 역할이지 실제 도메인 서비스는 아니므로 이름을 DataService로 변경한다.
- 이를 통해 repository간 중복 메서드를 제거한다.

Starlight258 committed:
feat: reservationStatus 필드 제거

- createdAt으로만 예약 확정자를 판단하는 로직을 구현한다.

Starlight258 committed:
refactor: equals & hashcode id로 판별하도록 정의

Starlight258 committed:
fix: 예약 조회시 예약 슬롯으로 조회되는 문제 해결

Starlight258 committed:
test: 테스트 버그 수정

Starlight258 committed:
refactor: 사용하지 않은 코드 제거

Starlight258 committed:
refactor: 메서드명 수정

Starlight258 committed:
feat: 예약 대기가 없을 경우 예약 slot 제거 기능 구현

Starlight258 committed:
feat: 예외 반환 클래스 수정

Starlight258 committed:
fix: 예약 제거되지 않은 문제 해결

Starlight258 committed:
feat: unique key 제약조건 추가

Starlight258 committed:
feat: 예약 제거시 예약 슬롯이 더이상 예약이 존재하지 않은 경우 제거 기능 구현

Starlight258 committed:
refactor: 사용하지 않은 메서드 제거 및 연관관계 메서드 재정립

- 사용하지 않은 메서드이거나 중복 메서드의 경우 제거한다.
- 연관관계 메서드를 진입점 중심으로 모은다.

Starlight258 committed:
refactor: 변수 및 메서드명 수정

Starlight258 committed:
feat: LocalDate.now()에 대해 clock 설정하여 테스트 용이성을 높인다.

Starlight258 committed:
refactor: 메서드 이름 변경 및 순서 조정

Starlight258 committed:
refactor: resolver 클래스명 수정

Starlight258 committed:
refactor: 메서드명 수정

Starlight258 committed:
refactor: ReservationService 책임 분리

- ReservationService가 확정 예약과 대기 예약 모두를 관리하므로, 서비스를 나누어 책임을 적절히 분리한다.

Starlight258 committed:
refactor: 서비스 dto 분리 및 dto명 수정

Starlight258 committed:
feat: 예약 관리 자료구조 List->Set 자료구조로 변경

- createdAt으로 확정 예약을 판단하므로 성능을 위해 HashSet으로 변경한다.

Starlight258 committed:
refactor: 변수명,클래스명 수정

Starlight258 committed:
fix: admin 권한이 필요한 api 권한 부여

- admin 권한이 필요한 api가 있다면 변경한다.

Starlight258 committed:
docs: api 문서 수정

Starlight258 committed:
fix: 불필요한 연관관계 설정 제거

Starlight258 committed:
feat: 내 예약 조회시 memberId 검증 추가

Starlight258 committed:
feat: 예외 타입 수정

Starlight258 committed:
feat: 예약 관리 자료구조 Set->List로 변경

- 프록시로 인한 객체 참조 문제를 해결한다.

Starlight258 committed:
refactor: 메서드 매개변수 수정

Starlight258 committed:
refactor: 메서드명 수정 및 순서 변경

Starlight258 committed:
fix: 불필요한 예약 대기 중 슬롯 삭제 로직 제거

- 예약 대기 중 슬롯을 삭제할 일이 없으므로 슬록 삭제 로직을 제거한다.

Starlight258 committed:
fix: 슬롯이 비어있을 경우 정상적으로 제거되도록 수정

- 예약이 삭제된 직후이므로 슬롯과 관련된 예약이 비어있을 경우 슬롯을 삭제한다.

Starlight258 committed:
refactor: 메서드명 수정

Starlight258 committed:
test: 예약 대기 Application 테스트 작성

Starlight258 committed:
refactor: reservationRepository 메서드명 수정 및 불필요한 jpql 제거

Starlight258 committed:
test: reservationRepository 테스트 작성

Starlight258 committed:
feat: 제거시 양방향 연관관계일 경우 관련 엔티티도 갱신

Starlight258 committed:
fix: 필요없는 검증 제거

Starlight258 committed:
fix: 필요없는 쿼리 메서드 제거

Starlight258 committed:
refactor: 메서드 파라미터를 쿼리 객체로 바인딩

Starlight258 committed:
test: 테스트 추가 및 메서드명 수정

Starlight258 committed:
feat: roleInterceptor에서 handlerMethod 조건부터 검사하도록 수정

Starlight258 committed:
feat: 인터셉터에서 user role까지 검증하도록 수정

Starlight258 committed:
refactor: 메서드명 수정

Starlight258 committed:
test: 테스트 보완

Starlight258 committed:
feat: @Transactional 어노테이션 서비스에 추가아혀 일관성을 높임

Starlight258 committed:
refactor: 사용하지 않은 클래스 삭제

Starlight258 committed:
refactor: 사용하지 않은 메서드 및 주석 제거

Starlight258 committed:
refactor: 권한별 controller 분리

Starlight258 committed:
refactor: now -> reservationDateTime으로 변경

Starlight258 committed:
feat: api endpoint 수정 및 메서드명 변경

Starlight258 committed:
feat: @OrderBy를 정의하여 컬렉션 로드시 자동으로 정렬되는 기능 추가

Starlight258 commented on src/main/java/roomescape/reservationslot/domain/ReservationSlot.java:
```diff
@@ -0,0 +1,149 @@
+    @JoinColumn(name = "theme_id", nullable = false)
+    private Theme theme;
+
+    @OneToMany(mappedBy = "reservationSlot", cascade = CascadeType.ALL, fetch = FetchType.LAZY, orphanRemoval = true)
+    @OrderBy("createdAt ASC")
+    private List<Reservation> reservations = new ArrayList<>();
```
### 일대다 양방향
- 이번에 다대다 테이블을 일대다, 다대일로 쪼개는 과정에서 `@OneToMany`를 이용해 양방향 관계로 설정해주었습니다.
- 그런데 `Reservation`은 외래키의 주인이 아니므로 저장하거나 삭제할 때 db에 바로 반영이 되지 않습니다. 따라서 연관된 `ReservationSlot`에 대한 관계를 매번 갱신해주었습니다. 
- 이 과정에서 기능을 구현하면서 "`일대다 양방향이 쉽지 않다`" 라는 고민을 하게 되었습니다.
- 실무에서도 양방향을 많이 쓰시는지 궁금해요!

summerlunaa replied:
> 실무에서도 양방향을 많이 쓰시는지 궁금해요!

너무 팀바팀일 것 같긴 하지만 제한적으로 사용하고는 있습니다. (상품 - 상품 이미지)나 (상품 - 상품 옵션) 등의 관계에서 사용하고 있어요. 개인적으로는 다루기 어려워서 별로 선호하지는 않습니다만 잘 사용한다면 아주 편하게 연관 데이터들을 다룰 수 있는 것 같아요. 장점이 많은 만큼 단점도 큰 설계 방식이라 잘 사용할 수 있도록 공부가 필요하고, 여러 경험을 통해 양방향 관계를 적용하기 적절한 케이스인지 아닌지를 구분하는 능력을 기르는 게 중요한 것 같습니다. '이럴 때 쓰세요'라고 제가 정해드릴 수는 없고 여러 시행착오와 경험을 통해 스스로 기준을 쌓아나갈 수 있다고 생각합니다 :) 이번 기회에 잘 적용해보신 것 같아요!

Starlight258 replied:
감사합니다! `@OneToMany`를 통해 참조 데이터를 쉽게 관리할 수 있었지만, 양방향 참조로 인해 신경써야할 부분이 많았어요. tradeoff를 고려해서 잘 설계할 수 있도록 열심히 공부하겠습니다!

Starlight258 commented on src/main/java/roomescape/member/presentation/controller/regular/RegularMemberController.java:
```diff
@@ -0,0 +1,41 @@
+import roomescape.member.domain.MemberRole;
+
+@RestController
+public class RegularMemberController {
```
### `@WebMvcTest`
<img width="567" alt="image" src="https://github.com/user-attachments/assets/4aa9bfcc-ddd8-4261-b051-1c8324750b6f" />

미션 목표 중 하나는 "`@DataJpaTest,` `@WebMvcTest `등을 사용하여 특정 계층의 테스트를 수행할 수 있다." 입니다.
따라서 컨트롤러를 테스트하는 단위 테스트인 `@WebMvcTest`에 대한 테스트를 작성하는 것에 고민이 들었습니다.

제 생각엔 컨트롤러에 비즈니스 로직이 없는데 테스트할 필요가 없다고 생각해서 작성하지 않았는데, 파랑은 컨트롤러 단위 테스트 많이 작성하시나요? 궁금합니다!

summerlunaa replied:
ㅎㅎㅎ 사실 저도 컨트롤러 테스트는 잘 작성하진 않아요. 말씀하신 것처럼 컨트롤러에는 이렇다할 로직이 없어서요. 인수테스트를 잘 작성하셨고, 대부분의 컨트롤러가 단순 service 호출만 하고 있다면 controller 테스트는 생략할 수도 있다고 생각합니다 :)
단위 테스트도 단순히 다른 클래스의 메서드를 호출만 하는 한 줄짜리 코드라면 생략할 수 있듯이 컨트롤러 테스트도 마찬가지인 것 같아요.

Starlight258 replied:
감사합니다! 역시 무조건은 없군요! 
테스트를 작성하고, 수정하는데에도 비용이 드니 필수적인 부분부터 우선순위에 따라 작성하는 것이 좋을 것 같아요~

Starlight258 commented:
파랑 안녕하세요!
피드백 답변을 통해 많이 배우고 성장하였습니다. 좋은 질문과 피드백 감사드려요! 😁

설계를 수정하면서 버그를 해결하고, 부족한 테스트를 작성하느라 늦게 답변 보낸 점 죄송합니다 🙇‍♀️
리뷰 감사드립니다!

Starlight258 committed:
fix: /login/check api 권한 수정

- guest 권한으로 수정

Starlight258 committed:
test: /login/check 테스트 guestTest로 이동

Starlight258 committed:
docs: api 문서 수정

summerlunaa reviewed (approved):
밍트 안녕하세요!
다음 미션과 병행하느라 시간이 많이 없었을텐데 리뷰 반영하느라 고생하셨습니다. 남겨주신 질문에도 답변 드렸어요.
방탈출 예약 대기 미션은 여기서 마무리 하겠습니다! 이제 정말 마지막 미션만 남았네요ㅎㅎㅎ 너무 고생하셨고 남은 우테코 생활도 응원하겠습니다. 수고하셨습니다 👏

Starlight258 commented:
리뷰 감사합니다 파랑 
피드백으로 인해 스스로 돌아보고 많이 성장할 수 있었어요!
감사했습니다 !! 🙇‍♀️
